uk.co.compucorp.civicrm.hrcore/CRM/HRCore/Helper/Menu/Parser.php

```diff
@@ -0,0 +1,85 @@
+<?php
+
+use CRM_HRCore_Menu_Item as MenuItem;
+
+class CRM_HRCore_Helper_Menu_Parser {
+
+  /**
+   * Builds the navigation menu tree and returns in a format expected
+   * by Civi in order to generate the navigation menu. It accepts an
+   * array of MenuItem objects
+   *
+   * @param array $menuObjects
+   *   An array of MenuItem Objects
+   *
+   * @return array
+   *  Sample return format:
+   *  [
+   *    'attributes' =>
+   *      [
+   *        'label' => 'Components',
+   *         'url' => 'civicrm/bla',
+   *         'permission' => 'sample permission',
+   *         'operator' => 'AND',
+   *      ],
+   *   ],
+   */
+  public static function getNavigationTree(array $menuObjects) {
```

Comment:
I thought this method might just be used in probably one place and its more of an helper method and there won't be need to mock it in tests, thats why I made it static.

The reason why its not a single menu instance to hold the entire menu instance is because, the menu items does not have a single parent, The top level navigation items Home, Staff, Leave etc are represented as menu item objects that contain their children within them since these top level menu items themselves have no parent, that is why they are supplied as an an array of parent (top level navigation menus) menu items.
What do you think?

Reply:
I think that, in general, we should start adopting the practice of always using instance methods by default and only use static one in situations where it is really necessary and/or make sense for the implementation. I want to see us applying more OO techniques in our codebase and using static methods kind goes against that.

The idea of having a class representing the menu goes along with this goal of making the codebase more OO. There are multiple ways of representing that. One way is as KnpMenu does, which is using an "blank" menu item instance which will only be used to keep a list of children (Home, Staff, Leave, etc) and each child can have its own list of children. Another way is to do something we have in this [laravel-menu library](https://github.com/lavary/laravel-menu), where there's a Menu class, which represents the entire menu, and an Item class, which represents the items in these menus.

The bottom line is, we need to represent the Menu in an OO format and we cannot do that with an array, because arrays are not objects and can contain any type of values.